// TODO nf-core: If in doubt look at other nf-core/modules to see how we are doing things! :)
//               https://github.com/nf-core/modules/tree/master/modules/nf-core/
//               You can also ask for help via your pull request or on the #modules channel on the nf-core Slack workspace:
//               https://nf-co.re/join
// TODO nf-core: A module file SHOULD only define input and output files as command-line parameters.
//               All other parameters MUST be provided using the "task.ext" directive, see here:
//               https://www.nextflow.io/docs/latest/process.html#ext
//               where "task.ext" is a string.
//               Any parameters that need to be evaluated in the context of a particular sample
//               e.g. single-end/paired-end data MUST also be defined and evaluated appropriately.
// TODO nf-core: Software that can be piped together SHOULD be added to separate module files
//               unless there is a run-time, storage advantage in implementing in this way
//               e.g. it's ok to have a single module for bwa to output BAM instead of SAM:
//                 bwa mem | samtools view -B -T ref.fasta
// TODO nf-core: Optional inputs are not currently supported by Nextflow. However, using an empty
//               list (`[]`) instead of a file can be used to work around this issue.

process QUERYNATOR_CGIAPI {
    tag "$meta.id"
    label 'process_low'

    conda "bioconda::querynator=0.1.3"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/querynator:0.1.3':
        'quay.io/biocontainers/querynator:0.1.3' }"
    
    input:

    tuple val(meta), path(mutations), path(cnas), path(translocations), val(cancer), val(genome), val(token), val(email)

    output:
    
    publishDir "${meta.id}", mode: 'copy', pattern: "*"

    path("${meta.id}.cgi_results.zip"), emit: zip
    path("${meta.id}.cgi_results"), emit: results    
    path("${meta.id}.cgi_results/drug_prescription.tsv"), emit: drug_tsv
    path("${meta.id}.cgi_results/input01.tsv"), emit: input_tsv
    path("${meta.id}.cgi_results/mutation_analysis.tsv"), emit: mutation_tsv
    path("${meta.id}.cgi_results/metadata.txt"), emit: metadata_txt

    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    // TODO nf-core: Where possible, a command MUST be provided to obtain the version number of the software e.g. 1.10
    //               If the software is unable to output a version number on the command-line then it can be manually specified
    //               e.g. https://github.com/nf-core/modules/blob/master/modules/nf-core/homer/annotatepeaks/main.nf
    //               Each software used MUST provide the software name and version number in the YAML version file (versions.yml)
    // TODO nf-core: It MUST be possible to pass additional parameters to the tool as a command-line string via the "task.ext.args" directive
    // TODO nf-core: If the tool supports multi-threading then you MUST provide the appropriate parameter
    //               using the Nextflow "task" variable e.g. "--threads $task.cpus"
    // TODO nf-core: Please replace the example samtools command below with your module's command
    // TODO nf-core: Please indent the command appropriately (4 spaces!!) to help with readability ;)

    // if parameter was set: use the flag with the file, if not: command on console is ''

    def mutations_file = mutations ? "--mutations ${mutations}" : "" 
    def translocation_file = translocations ? "--translocations ${translocations}" : ''
    def cnas_file = cnas ? "--cnas ${cnas}" : ''
    
    """
    querynator query-api-cgi \
        $mutations_file \
        $translocation_file \
        $cnas_file \
        --output $prefix \
        --cancer $cancer \
        --genome $genome \
        --token $token \
        --email $email \
    

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        querynator: \$(echo \$(querynator --version 2>&1) | sed 's/^.*querynator //; s/Using.*\$//' ))
    END_VERSIONS
    """
}
